### PR TITLE
Use Haskell variables for passing values.

### PR DESCRIPTION
**Pull request by MarcFontaine**
This is a combination of several commits. (They were squashed into one big commit during an elaborate re-base).
Most changes are related to the way filenames work in `cardano-testnet`.

Currently the `cardano-testnet` uses filenames (Strings) to reference objects like keys, certs etc.
Two locations in the code refer to the same object by "magically" using the same filename.
This means that the data flow is implicit and hard to follow.
Old scheme:
```
do
cli_call_1 "filenameX" "filenameY"
cli_call_2 "filenameX" "filenameZ"
cli_call_3 "filenameA" "filenameB"
```
This PR changes the design as follows:
* File names are stored in an opaque `File` datatype (which wraps the file path).
* A wrapper is used for CLI calls. When a CLI call creates a file the wrapper returns the corresponding `File` resource.
* While CLI calls have type `IO ()` the wrapper has type `IO (File,..,File).
* CLI calls don't construct magic Filenames, but instead use the `File` resource that was created earlier.
The new scheme is
```
do
(file1, file2) <- cli_call1 arg1 arg2     -- cli_call1 creates file resources file1 and file2
(fileX, fileY) <- cli_call2 file1 file2    -- cli_call2 uses file resourcs file1 and file2
```
Full transformation to new scheme for filenames is still WIP.

### CODE REVIEW

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
There are a couple issues with this PR that need to be addressed before it is reviewed:
- There are several left over comments that should not be there
- There is a single commit for a relatively large diff

You need to remove comments that are no longer required and break this PR up in to several commits with clear commit messages describing the changes in the given commit. This makes it easier for a reviewer to understand the particular changes you have made.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:8]: Why do we need `no-unused-matches` and `no-unused-top-binds`?

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:221]: Can you elaborate why you mean be "dead code".  There are a number of such references.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:278]: Redundant parentheses.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:363]: I think adding a newline here adds clarity.

File: cardano-testnet/test/Test/FoldBlocks.hs
Comment (newhoggy) [inline cardano-testnet/test/Test/FoldBlocks.hs:43]: I know this is old code, but can you find out why?

File: cardano-testnet/test/Test/FoldBlocks.hs
Comment (newhoggy) [inline cardano-testnet/test/Test/FoldBlocks.hs:40]: > outside of the package directory !!

Yeah, that's what `base` and `getProjectBase` means.  Is there a way to make this clearer without adding a comment?

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:141]: Looks a bit crowded.  Can you spread them out with newlines?

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:130]: Double space between `)  (` could be one line.

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:75]: I'm curious about the comment.  We shouldn't have to care where this data is used.  For example new tests might be added that make use of this.

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:79]: I'm partial to using something more type safe than `FilePath` for each type of file to make usage information easier to find but it doesn't have to be this PR.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:396]: Can the typing be done by type applications instead of type ascriptions?

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:447]: These were translated directly from the `mkfiles.sh` script this was based on.  Those scripts were designed so that the update proposals can be submitted manually.

We should consider what we want to do with this in the Haskell code.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:61]: Redundant parentheses.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:56]: This will need to be MacOS and Windows friendly.  MacOS filenames are case insensitive.  Windows filenames are case-insensitive and many characters have special meaning.

Perhaps we can call it `Kebab` because it's using kebab-casing and we lock it down do only support lower case alpha-numerics.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:138]: Nice!  Can you add a newline above the `newtype`?

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:74]: The `KeyGen` type alias seems less helpful.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:160]: I think I rather just use `fst` and `snd` and `unFile`.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:151]: Commented out code.

**Comment by newhoggy** [comment]
I agree with @Jimbo4350 regarding breaking up the code into multiple commits.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:8]: The refactor uses Haskell variables for file resources.
The CLI wrapper functions return those resources.
However the refactor is still WIP some resources are defined but not yet used.
Once everything is don't the warnings should be turned on again.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:221]: Some parts of the code code call `cardano-cli` for example to create a transaction.
However that transaction is never ever transmitted to the testnet or reused.
Also the code is not tested, all the tests pass when the code is disabled.
`cardanoTestnetByronGovernance` creates genesis key, delegation certs and delegation keys,
however this are are not used.

File: cardano-testnet/test/Test/FoldBlocks.hs
Comment (MarcFontaine) [reply cardano-testnet/test/Test/FoldBlocks.hs:43]: I only stumbled upon this because `(tempAbsBasePath' <> "/")` looks strange.
but as the comment says the tests fail without the trailing `<> "/"`.
I'm in the course of reordering the various file paths and this will probably be removed anyway.

File: cardano-testnet/test/Test/FoldBlocks.hs
Comment (MarcFontaine) [reply cardano-testnet/test/Test/FoldBlocks.hs:40]: That comment is a TODO for me.
I think it is better to move copies of those file to the package itself
or maybe there is yet an other way to do this.

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Runtime.hs:141]: ups yes some blank lines got messed up during the rebase

File: cardano-testnet/src/Testnet/Util/Runtime.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Runtime.hs:75]: I'm about to change the code so I naturally need to know where something is defined and where it is used.
I think I also found stuff that is defined but not not used.
IMHO most of the time, it is not a good idea to add record fields for data that is not used.
Just because the data might be used in the future is not a good reason.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:396]: Thanks for spotting, those types can be inferred and annotation are superfluous.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:421]: There is more to be done here..its dead code inside dead code.
`forM [1 .. numBftNodes]` keys are generated but only the first is used.

File: cardano-testnet/src/Testnet/Cardano.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:447]: Good information.
I should have a look at `mkfiles.sh`. Is it still available ? Is it still in use ?
I'll change that comment to a TODO:
There should be a test that submits the update proposals and checks that they are activated.
It was not clear to me, if that test had been removed or disabled, or if it had never existed.

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:160]: Interesting. I guess I had written the type signature `fst` copied it to a `snd` line
and then edited that line to become the function definition...

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:151]: Is there a simple function in the test frame-work that asserts that a file exists ?
Last time I looked at it there was only one which checks for a list of files.

**Comment by MarcFontaine** [comment]
> I agree with @Jimbo4350 regarding breaking up the code into multiple commits.

I break up my code into multiple commits anyway.
However, here it was impossible to rebase it as such, so I squashed the PR into one commit.
Overall this PR is very simple and not extremely long.
Did splitting up the PR again it make easier to review for you ?

File: cardano-testnet/src/Testnet/Util/Cli.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:74]: I do not know.. I looked up the code `KeyGen` is used five times.

File: cardano-testnet/src/Testnet/Conf.hs
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Conf.hs:27]: I can 100% get behind this simplification :+1:

File: cardano-testnet/src/Testnet/Babbage.hs
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Babbage.hs:133]: Actually, it got called in one of the tests and
the error message contained the exact Module and line number.
The problem is in the `TestnetRuntime` type, which can either fit `babbage` or `cardano` but not both eras.
(That whole block of code is to be removed in a followup PR)

**Comment by Jimbo4350** [comment]
> @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.

@MarcFontaine one the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the types of PRs that I'm looking for.

**Comment by MarcFontaine** [comment]
> > @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.
> 
> @MarcFontaine on the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the type of PRs that I'm looking for.

The job of hlint is best done by CI.
How about this:
1) If there are no technical issues with a PR it gets a review with an approval
2) If it also passes CI (including of cause the linter) it gets merged.
3) We happily discuss "other style issues" as long as it is fun and does not kill productivity 
4) "other style issues" don"t hold back a PR.

**Comment by Jimbo4350** [comment]
> > > @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.
> > 
> > 
> > @MarcFontaine on the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the type of PRs that I'm looking for.
> 
> The job of hlint is best done by CI. How about this:
> 
>     1. If there are no technical issues with a PR it gets a review with an approval
> 
>     2. If it also passes CI (including of cause the linter) it gets merged.
> 
>     3. We happily discuss "other style issues" as long as it is fun and does not kill productivity
> 
>     4. "other style issues" don"t hold back a PR.

No. This PR can still be broken up and PRs need to be of a certain standard before merging. What you are suggesting will create more churn and make the code less readable. As I said I will demonstrate what I am looking for.

**Comment by MarcFontaine** [comment]
> No. This PR can still be broken up and PRs need to be of a certain standard, both technical and stylistic, before merging. What you are suggesting will create more churn and make the code less readable. As I said I will demonstrate what I am looking for.

The orginal code has been written two years ago and nobody would care about improving it or moving it forward.
I open a PR and try to propose some changes and suddenly people start to merge PR after PR that change right that code. Honestly, I do not care to much about demonstrating anything to anybody..

**Pull request closed without merging**
